File: scripts/setup_appwrite.py
#!/usr/bin/env python3
"""
FitTrack Appwrite Provisioning Script
Automatically creates the FitTrack database, 7 collections, attributes, indexes, and storage bucket in your Appwrite project.
"""

import sys
import json
import time
import urllib.request
import urllib.error

# Ensure immediate unbuffered console logging
if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(line_buffering=True)

DEFAULT_ENDPOINT = "https://cloud.appwrite.io/v1"
DEFAULT_DATABASE_ID = "fittrack"
DEFAULT_BUCKET_ID = "progress-photos"

def api_request(url, method="GET", headers=None, data=None):
    if headers is None:
        headers = {}
    req = urllib.request.Request(url, method=method, headers=headers)
    if data is not None:
        req.data = json.dumps(data).encode("utf-8")
        req.add_header("Content-Type", "application/json")
    try:
        with urllib.request.urlopen(req) as response:
            body = response.read().decode("utf-8")
            return json.loads(body) if body else {}
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8")
        try:
            err_json = json.loads(body)
        except Exception:
            err_json = {"message": body}
        return {"_error": True, "status": e.code, "error": err_json}
    except Exception as e:
        return {"_error": True, "status": 0, "error": str(e)}

def wait_for_attributes(endpoint, headers, database_id, collection_id):
    """Wait for all collection attributes to finish processing in Appwrite."""
    url = f"{endpoint}/databases/{database_id}/collections/{collection_id}/attributes"
    for _ in range(30):
        res = api_request(url, headers=headers)
        if res.get("_error"):
            time.sleep(1)
            continue
        attrs = res.get("attributes", [])
        if attrs and all(a.get("status") == "available" for a in attrs):
            return True
        time.sleep(1)
    return True

def setup_appwrite(endpoint, project_id, api_key, database_id=DEFAULT_DATABASE_ID, bucket_id=DEFAULT_BUCKET_ID):
    headers = {
        "X-Appwrite-Project": project_id,
        "X-Appwrite-Key": api_key,
    }

    print(f"\n==========================================")
    print(f" Connecting to Appwrite: {endpoint}")
    print(f" Project ID: {project_id}")
    print(f" Database ID: {database_id}")
    print(f" Bucket ID: {bucket_id}")
    print(f"==========================================\n")

    # 1. Create or verify Database
    print(f"[1/4] Checking Database '{database_id}'...")
    db_url = f"{endpoint}/databases/{database_id}"
    db_res = api_request(db_url, headers=headers)
    if db_res.get("_error") and db_res.get("status") == 404:
        print(f"  Creating database '{database_id}'...")
        create_db_res = api_request(
            f"{endpoint}/databases",
            method="POST",
            headers=headers,
            data={"databaseId": database_id, "name": "FitTrack Database"}
        )
        if create_db_res.get("_error"):
            print(f"  Error creating database: {create_db_res['error']}")
            return False
        print("  Database created successfully.")
    elif not db_res.get("_error"):
        print("  Database exists.")
    else:
        print(f"  Error checking database: {db_res['error']}")
        return False

    # 2. Define Collections & Attributes
    collections = {
        "profiles": {
            "name": "Profiles",
            "attributes": [
                {"type": "string", "key": "name", "size": 128, "required": False, "default": "Athlete"},
                {"type": "string", "key": "goal", "size": 64, "required": False, "default": "Build Muscle"},
                {"type": "float", "key": "height", "required": False, "default": 178.0},
                {"type": "float", "key": "current_weight", "required": False, "default": 74.2},
                {"type": "float", "key": "target_weight", "required": False, "default": 78.0},
                {"type": "string", "key": "preferred_reminder_time", "size": 16, "required": False, "default": "18:00:00"},
                {"type": "integer", "key": "workout_streak", "required": False, "default": 0},
                {"type": "integer", "key": "photo_streak", "required": False, "default": 0},
                {"type": "boolean", "key": "has_completed_onboarding", "required": False, "default": False},
            ],
            "indexes": []
        },
        "workouts": {
            "name": "Workouts",
            "attributes": [
                {"type": "string", "key": "user_id", "size": 64, "required": True},
                {"type": "string", "key": "title", "size": 128, "required": True, "default": "Workout Session"},
                {"type": "string", "key": "workout_date", "size": 64, "required": True},
                {"type": "integer", "key": "duration_minutes", "required": False, "default": 45},
                {"type": "string", "key": "notes", "size": 2048, "required": False, "default": ""},
            ],
            "indexes": [
                {"key": "idx_workouts_user_date", "type": "key", "attributes": ["user_id", "workout_date"], "orders": ["ASC", "DESC"]}
            ]
        },
        "workout_exercises": {
            "name": "Workout Exercises",
            "attributes": [
                {"type": "string", "key": "workout_id", "size": 64, "required": True},
                {"type": "string", "key": "user_id", "size": 64, "required": True},
                {"type": "string", "key": "exercise_name", "size": 128, "required": True},
                {"type": "integer", "key": "exercise_order", "required": False, "default": 0},
            ],
            "indexes": [
                {"key": "idx_exercises_workout", "type": "key", "attributes": ["workout_id", "exercise_order"], "orders": ["ASC", "ASC"]}
            ]
        },
        "workout_sets": {
            "name": "Workout Sets",
            "attributes": [
                {"type": "string", "key": "workout_id", "size": 64, "required": True},
                {"type": "string", "key": "workout_exercise_id", "size": 64, "required": True},
                {"type": "string", "key": "user_id", "size": 64, "required": True},
                {"type": "integer", "key": "set_number", "required": True},
                {"type": "float", "key": "weight", "required": False, "default": 0.0},
                {"type": "integer", "key": "reps", "required": False, "default": 0},
                {"type": "boolean", "key": "is_completed", "required": False, "default": True},
            ],
            "indexes": [
                {"key": "idx_sets_exercise", "type": "key", "attributes": ["workout_exercise_id", "set_number"], "orders": ["ASC", "ASC"]}
            ]
        },
        "personal_records": {
            "name": "Personal Records",
            "attributes": [
                {"type": "string", "key": "user_id", "size": 64, "required": True},
                {"type": "string", "key": "exercise_name", "size": 128, "required": True},
                {"type": "float", "key": "max_weight", "required": True},
                {"type": "integer", "key": "max_reps", "required": False, "default": 0},
                {"type": "string", "key": "achieved_at", "size": 64, "required": False, "default": ""},
                {"type": "string", "key": "workout_id", "size": 64, "required": False, "default": ""},
            ],
            "indexes": [
                {"key": "idx_pr_user_exercise", "type": "key", "attributes": ["user_id", "exercise_name"], "orders": ["ASC", "ASC"]}
            ]
        },
        "measurements": {
            "name": "Body Measurements",
            "attributes": [
                {"type": "string", "key": "user_id", "size": 64, "required": True},
                {"type": "string", "key": "measurement_type", "size": 64, "required": True},
                {"type": "float", "key": "value", "required": True},
                {"type": "string", "key": "unit", "size": 16, "required": False, "default": "kg"},
                {"type": "string", "key": "recorded_at", "size": 64, "required": True},
                {"type": "string", "key": "notes", "size": 1024, "required": False, "default": ""},
            ],
            "indexes": [
                {"key": "idx_measurements_user_type_date", "type": "key", "attributes": ["user_id", "measurement_type"], "orders": ["ASC", "ASC"]}
            ]
        },
        "progress_photos": {
            "name": "Progress Photos",
            "attributes": [
                {"type": "string", "key": "user_id", "size": 64, "required": True},
                {"type": "string", "key": "file_id", "size": 64, "required": True},
                {"type": "string", "key": "storage_path", "size": 128, "required": False, "default": ""},
                {"type": "string", "key": "pose", "size": 64, "required": False, "default": "Front"},
                {"type": "string", "key": "workout_id", "size": 64, "required": False, "default": ""},
                {"type": "float", "key": "weight_at_capture", "required": False, "default": 0.0},
                {"type": "string", "key": "notes", "size": 2048, "required": False, "default": ""},
                {"type": "string", "key": "created_at", "size": 64, "required": True},
            ],
            "indexes": [
                {"key": "idx_photos_user_date", "type": "key", "attributes": ["user_id", "created_at"], "orders": ["ASC", "DESC"]}
            ]
        }
    }

    # 3. Create Collections & Attributes
    print(f"\n[2/4] Setting up {len(collections)} Collections...")
    for col_id, col_info in collections.items():
        print(f"  Checking collection '{col_id}' ({col_info['name']})...")
        c_url = f"{endpoint}/databases/{database_id}/collections/{col_id}"
        c_res = api_request(c_url, headers=headers)

        if c_res.get("_error") and c_res.get("status") == 404:
            print(f"    Creating collection '{col_id}'...")
            create_c_res = api_request(
                f"{endpoint}/databases/{database_id}/collections",
                method="POST",
                headers=headers,
                data={
                    "collectionId": col_id,
                    "name": col_info["name"],
                    "permissions": [
                        'read("users")',
                        'create("users")',
                        'update("users")',
                        'delete("users")'
                    ],
                    "documentSecurity": True
                }
            )
            if create_c_res.get("_error"):
                print(f"    Error creating collection: {create_c_res['error']}")
                continue
        elif c_res.get("_error"):
            print(f"    Error checking collection: {c_res['error']}")
            continue

        # Create attributes
        for attr in col_info["attributes"]:
            attr_type = attr["type"]
            attr_key = attr["key"]
            attr_url = f"{endpoint}/databases/{database_id}/collections/{col_id}/attributes/{attr_type}"
            payload = {
                "key": attr_key,
                "required": attr.get("required", False),
            }
            if "size" in attr:
                payload["size"] = attr["size"]
            if not attr.get("required") and "default" in attr:
                payload["default"] = attr["default"]

            attr_res = api_request(attr_url, method="POST", headers=headers, data=payload)
            if attr_res.get("_error"):
                if attr_res.get("status") == 409:
                    pass  # Attribute already exists
                else:
                    print(f"      Attr '{attr_key}' notice: {attr_res['error'].get('message', attr_res['error'])}")
            else:
                print(f"      + Created attribute '{attr_key}' ({attr_type})")

    # Wait for attributes to become available before adding indexes
    print(f"\n[3/4] Creating Indexes...")
    for col_id, col_info in collections.items():
        if not col_info.get("indexes"):
            continue
        wait_for_attributes(endpoint, headers, database_id, col_id)
        for idx in col_info["indexes"]:
            idx_url = f"{endpoint}/databases/{database_id}/collections/{col_id}/indexes"
            idx_res = api_request(idx_url, method="POST", headers=headers, data=idx)
            if idx_res.get("_error"):
                if idx_res.get("status") == 409:
                    pass  # Index already exists
                else:
                    print(f"    Index '{idx['key']}' notice: {idx_res['error'].get('message', idx_res['error'])}")
            else:
                print(f"    + Created index '{idx['key']}' in '{col_id}'")

    # 4. Create Storage Bucket
    print(f"\n[4/4] Checking Storage Bucket '{bucket_id}'...")
    b_url = f"{endpoint}/storage/buckets/{bucket_id}"
    b_res = api_request(b_url, headers=headers)
    if b_res.get("_error") and b_res.get("status") == 404:
        print(f"  Creating bucket '{bucket_id}'...")
        create_b_res = api_request(
            f"{endpoint}/storage/buckets",
            method="POST",
            headers=headers,
            data={
                "bucketId": bucket_id,
                "name": "Progress Photos",
                "permissions": [
                    'read("users")',
                    'create("users")',
                    'update("users")',
                    'delete("users")'
                ],
                "fileSecurity": True,
                "enabled": True,
                "maximumFileSize": 15728640,
                "allowedFileExtensions": ["jpg", "jpeg", "png", "webp"],
                "compression": "none",
                "encryption": True,
                "antivirus": True
            }
        )
        if create_b_res.get("_error"):
            print(f"  Error creating bucket: {create_b_res['error']}")
        else:
            print("  Storage bucket created successfully.")
    elif not b_res.get("_error"):
        print("  Storage bucket exists.")
    else:
        print(f"  Error checking bucket: {b_res['error']}")

    print("\n==========================================")
    print(" [SUCCESS] Appwrite FitTrack Setup Complete!")
    print("==========================================\n")
    return True

if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("\nUsage:")
        print("  python scripts/setup_appwrite.py <PROJECT_ID> <API_KEY> [ENDPOINT]\n")
        print("Example:")
        print("  python scripts/setup_appwrite.py 65a1b2c3d4e5f6 key_abc123... https://cloud.appwrite.io/v1\n")
        sys.exit(1)

    project_id = sys.argv[1]
    api_key = sys.argv[2]
    endpoint = sys.argv[3] if len(sys.argv) > 3 else DEFAULT_ENDPOINT

    setup_appwrite(endpoint, project_id, api_key)

File: scripts/setup_appwrite.js
/**
 * FitTrack - Appwrite Database & Storage Setup Script
 *
 * Automatically creates the `fittrack` database, all 7 collections,
 * all attributes, indexes, and storage bucket on Appwrite Cloud or Self-Hosted.
 *
 * Usage:
 *   node scripts/setup_appwrite.js <YOUR_APPWRITE_API_KEY>
 *
 * Or set environment variable:
 *   APPWRITE_API_KEY=your_key node scripts/setup_appwrite.js
 */

const ENDPOINT = process.env.APPWRITE_ENDPOINT || 'https://sgp.cloud.appwrite.io/v1';
const PROJECT_ID = process.env.APPWRITE_PROJECT_ID || '6aac02f1002c53d0bc56';
const DATABASE_ID = process.env.APPWRITE_DATABASE_ID || 'fittrack';
const BUCKET_ID = process.env.APPWRITE_PHOTOS_BUCKET || 'progress-photos';
const API_KEY = process.argv[2] || process.env.APPWRITE_API_KEY;

if (!API_KEY) {
  console.log(`
================================================================
  FitTrack Appwrite Database Setup
================================================================
  Please provide your Appwrite API Key to provision collections.

  How to get an Appwrite API Key:
  1. Open Appwrite Console: https://cloud.appwrite.io/console/project/${PROJECT_ID}
  2. Go to: Overview > Integrations > API Keys (or Project Settings > API Keys)
  3. Click "Create API Key"
  4. Name: "FitTrack Setup", Expiration: None / Desired
  5. Scopes: Check "databases.write", "collections.write", "attributes.write", "indexes.write", "buckets.write"
  6. Copy the generated secret key and run:

     node scripts/setup_appwrite.js <YOUR_API_KEY>
================================================================
`);
  process.exit(1);
}

const headers = {
  'Content-Type': 'application/json',
  'X-Appwrite-Project': PROJECT_ID,
  'X-Appwrite-Key': API_KEY,
};

async function api(path, method = 'GET', body = null) {
  const options = { method, headers };
  if (body) options.body = JSON.stringify(body);
  const res = await fetch(`${ENDPOINT}${path}`, options);
  const text = await res.text();
  let data;
  try {
    data = JSON.parse(text);
  } catch {
    data = { raw: text };
  }
  return { status: res.status, ok: res.ok, data };
}

async function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

async function main() {
  console.log(`Connecting to Appwrite: ${ENDPOINT} (Project: ${PROJECT_ID})`);

  // 1. Create Database
  console.log(`\n[1/8] Ensuring database '${DATABASE_ID}' exists...`);
  const dbRes = await api('/databases', 'POST', {
    databaseId: DATABASE_ID,
    name: 'FitTrack Database',
    enabled: true,
  });
  if (dbRes.ok) {
    console.log(`  ✓ Database '${DATABASE_ID}' created.`);
  } else if (dbRes.status === 409) {
    console.log(`  ✓ Database '${DATABASE_ID}' already exists.`);
  } else {
    console.log(`  Note on database: ${dbRes.data.message || dbRes.status}`);
  }

  // 2. Collections definition
  const collections = [
    {
      id: 'profiles',
      name: 'User Profiles',
      attributes: [
        { key: 'name', type: 'string', size: 128, required: false, default: 'Athlete' },
        { key: 'goal', type: 'string', size: 64, required: false, default: 'Build Muscle' },
        { key: 'height', type: 'float', required: false, default: 178.0 },
        { key: 'current_weight', type: 'float', required: false, default: 74.2 },
        { key: 'target_weight', type: 'float', required: false, default: 78.0 },
        { key: 'preferred_reminder_time', type: 'string', size: 16, required: false, default: '18:00:00' },
        { key: 'workout_streak', type: 'integer', required: false, default: 0 },
        { key: 'photo_streak', type: 'integer', required: false, default: 0 },
        { key: 'has_completed_onboarding', type: 'boolean', required: false, default: false },
      ],
      indexes: [],
    },
    {
      id: 'workouts',
      name: 'Workouts',
      attributes: [
        { key: 'user_id', type: 'string', size: 64, required: true },
        { key: 'title', type: 'string', size: 128, required: true },
        { key: 'workout_date', type: 'string', size: 64, required: true },
        { key: 'duration_minutes', type: 'integer', required: false, default: 45 },
        { key: 'notes', type: 'string', size: 2048, required: false },
      ],
      indexes: [
        { key: 'idx_workouts_user_date', type: 'key', attributes: ['user_id', 'workout_date'], orders: ['ASC', 'DESC'] },
      ],
    },
    {
      id: 'workout_exercises',
      name: 'Workout Exercises',
      attributes: [
        { key: 'workout_id', type: 'string', size: 64, required: true },
        { key: 'user_id', type: 'string', size: 64, required: true },
        { key: 'exercise_name', type: 'string', size: 128, required: true },
        { key: 'exercise_order', type: 'integer', required: false, default: 0 },
        { key: 'notes', type: 'string', size: 1024, required: false },
      ],
      indexes: [
        { key: 'idx_exercises_workout', type: 'key', attributes: ['workout_id', 'exercise_order'], orders: ['ASC', 'ASC'] },
      ],
    },
    {
      id: 'workout_sets',
      name: 'Workout Sets',
      attributes: [
        { key: 'workout_id', type: 'string', size: 64, required: true },
        { key: 'workout_exercise_id', type: 'string', size: 64, required: true },
        { key: 'user_id', type: 'string', size: 64, required: true },
        { key: 'set_number', type: 'integer', required: true, default: 1 },
        { key: 'weight', type: 'float', required: false, default: 0.0 },
        { key: 'reps', type: 'integer', required: false, default: 0 },
        { key: 'is_completed', type: 'boolean', required: false, default: true },
      ],
      indexes: [
        { key: 'idx_sets_exercise', type: 'key', attributes: ['workout_exercise_id', 'set_number'], orders: ['ASC', 'ASC'] },
      ],
    },
    {
      id: 'personal_records',
      name: 'Personal Records',
      attributes: [
        { key: 'user_id', type: 'string', size: 64, required: true },
        { key: 'exercise_name', type: 'string', size: 128, required: true },
        { key: 'max_weight', type: 'float', required: true, default: 0.0 },
        { key: 'max_reps', type: 'integer', required: false, default: 0 },
        { key: 'achieved_at', type: 'string', size: 64, required: false },
        { key: 'workout_id', type: 'string', size: 64, required: false },
      ],
      indexes: [
        { key: 'idx_pr_user_exercise', type: 'key', attributes: ['user_id', 'exercise_name'], orders: ['ASC', 'ASC'] },
      ],
    },
    {
      id: 'measurements',
      name: 'Body Measurements',
      attributes: [
        { key: 'user_id', type: 'string', size: 64, required: true },
        { key: 'measurement_type', type: 'string', size: 64, required: true },
        { key: 'value', type: 'float', required: true },
        { key: 'unit', type: 'string', size: 16, required: false, default: 'kg' },
        { key: 'recorded_at', type: 'string', size: 64, required: true },
        { key: 'notes', type: 'string', size: 1024, required: false },
      ],
      indexes: [
        { key: 'idx_measurements_user_type_date', type: 'key', attributes: ['user_id', 'measurement_type', 'recorded_at'], orders: ['ASC', 'ASC', 'DESC'] },
      ],
    },
    {
      id: 'progress_photos',
      name: 'Progress Photos',
      attributes: [
        { key: 'user_id', type: 'string', size: 64, required: true },
        { key: 'file_id', type: 'string', size: 64, required: true },
        { key: 'storage_path', type: 'string', size: 128, required: false },
        { key: 'pose', type: 'string', size: 64, required: false, default: 'Front' },
        { key: 'workout_id', type: 'string', size: 64, required: false },
        { key: 'weight_at_capture', type: 'float', required: false },
        { key: 'notes', type: 'string', size: 2048, required: false },
        { key: 'day_number', type: 'integer', required: false, default: 1 },
        { key: 'created_at', type: 'string', size: 64, required: true },
      ],
      indexes: [
        { key: 'idx_photos_user_date', type: 'key', attributes: ['user_id', 'created_at'], orders: ['ASC', 'DESC'] },
        { key: 'idx_photos_user_day', type: 'key', attributes: ['user_id', 'day_number'], orders: ['ASC', 'ASC'] },
      ],
    },
  ];

  for (let i = 0; i < collections.length; i++) {
    const col = collections[i];
    console.log(`\n[${i + 2}/8] Creating collection '${col.id}' (${col.name})...`);

    const colRes = await api(`/databases/${DATABASE_ID}/collections`, 'POST', {
      collectionId: col.id,
      name: col.name,
      permissions: ['read("any")', 'create("any")', 'update("any")', 'delete("any")'],
      documentSecurity: false,
      enabled: true,
    });

    if (colRes.ok) {
      console.log(`  ✓ Collection '${col.id}' created.`);
    } else if (colRes.status === 409) {
      console.log(`  ✓ Collection '${col.id}' already exists.`);
    } else {
      console.log(`  Note on collection: ${colRes.data.message || colRes.status}`);
    }

    // Create attributes
    for (const attr of col.attributes) {
      const path = `/databases/${DATABASE_ID}/collections/${col.id}/attributes/${attr.type}`;
      const payload = {
        key: attr.key,
        required: attr.required,
      };
      if (attr.type === 'string') payload.size = attr.size || 128;
      if (attr.default !== undefined) payload.default = attr.default;

      const attrRes = await api(path, 'POST', payload);
      if (attrRes.ok) {
        console.log(`    + Attribute '${attr.key}' (${attr.type}) created.`);
      } else if (attrRes.status === 409) {
        console.log(`    + Attribute '${attr.key}' exists.`);
      } else {
        console.log(`    ! Attribute '${attr.key}': ${attrRes.data.message || attrRes.status}`);
      }
      await sleep(150);
    }

    // Wait for attributes to become available before adding indexes
    if (col.indexes.length > 0) {
      await sleep(1000);
      for (const idx of col.indexes) {
        const idxRes = await api(`/databases/${DATABASE_ID}/collections/${col.id}/indexes`, 'POST', {
          key: idx.key,
          type: idx.type,
          attributes: idx.attributes,
          orders: idx.orders,
        });
        if (idxRes.ok) {
          console.log(`    # Index '${idx.key}' created.`);
        } else if (idxRes.status === 409) {
          console.log(`    # Index '${idx.key}' exists.`);
        } else {
          console.log(`    ! Index '${idx.key}': ${idxRes.data.message || idxRes.status}`);
        }
        await sleep(150);
      }
    }
  }

  // 8. Storage Bucket
  console.log(`\n[8/8] Ensuring storage bucket '${BUCKET_ID}' exists...`);
  const bucketRes = await api('/storage/buckets', 'POST', {
    bucketId: BUCKET_ID,
    name: 'Progress Photos',
    permissions: ['read("any")', 'create("any")', 'update("any")', 'delete("any")'],
    fileSecurity: false,
    enabled: true,
    maximumFileSize: 15 * 1024 * 1024,
    allowedFileExtensions: ['jpg', 'jpeg', 'png', 'webp', 'json'],
  });

  if (bucketRes.ok) {
    console.log(`  ✓ Storage bucket '${BUCKET_ID}' created.`);
  } else if (bucketRes.status === 409) {
    console.log(`  ✓ Storage bucket '${BUCKET_ID}' already exists.`);
  } else {
    console.log(`  Note on storage bucket: ${bucketRes.data.message || bucketRes.status}`);
  }

  console.log(`
================================================================
  ✓ All FitTrack database collections and storage initialized!
================================================================
`);
}

main().catch((err) => {
  console.error('Fatal error during setup:', err);
  process.exit(1);
});
